In [26]:
# Run nbconvert from a notebook cell using the shell. Prefix with "!" so it's executed as a shell command.
# Use forward slashes to avoid backslash-escape issues on Windows paths.
!jupyter nbconvert --to html "C:/xampp/htdocs/ml_personalNGroup/Group/chalenge2/notebooks/02-feature-engineering.ipynb"

[NbConvertApp] Converting notebook C:/xampp/htdocs/ml_personalNGroup/Group/chalenge2/notebooks/02-feature-engineering.ipynb to html
[NbConvertApp] Writing 308387 bytes to C:\xampp\htdocs\ml_personalNGroup\Group\chalenge2\notebooks\02-feature-engineering.html


## 1. Tải dữ liệu

In [1]:
import pandas as pd
df_train = pd.read_csv('../data/01_raw/train.csv')
df_test = pd.read_csv('../data/01_raw/test.csv')

## 2. Kiểm tra dữ liệu từ train

In [2]:
df_train.head()
df_train.info()
# Tạo DataFrame mô tả
df_train.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

## Xử lí giá trị thiếu

## Cột có nhiều giá trị thiếu nhưng ý nghĩa là không có

Ví dụ các cột như: Alley, FireplaceQu, PoolQC, Fence, MiscFeature, GarageType, GarageFinish, GarageQual, GarageCond, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, MasVnrType

In [28]:
import pandas as pd
from datetime import datetime
import os

# Giả sử bạn đã xử lý missing values như sau:
cols_none = [
    'Alley', 'FireplaceQu', 'PoolQC', 'Fence', 'MiscFeature',
    'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond',
    'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'MasVnrType'
]
# Chỉ điền giá trị thiếu cho các cột còn tồn tại
for col in cols_none:
    if col in df_train.columns:
        df_train[col] = df_train[col].fillna('None')
    if col in df_test.columns:
        df_test[col] = df_test[col].fillna('None')

# --- Ghi lại nhật ký xử lý ---
excel_file = "../reports/featureEngineering_summary.xlsx"
now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

# Kết luận
note = (
    "Đã xử lý các cột có giá trị thiếu dạng None (các cột liên quan đến thông tin Garage, Basement, "
    "và tường veneer). Sau khi xử lý, không còn giá trị NaN trong các cột này."
)

# Tạo dataframe lưu log
log_entry = pd.DataFrame({
    "Thời gian": [now],
    "Hành động": ["Xử lý Missing Values (fillna('None'))"],
    "Kết luận": [note]
})

# Nếu file chưa tồn tại, tạo mới
if not os.path.exists(excel_file):
    log_entry.to_excel(excel_file, index=False)
else:
    # Nếu đã có file, append thêm dòng mới
    old_log = pd.read_excel(excel_file)
    updated_log = pd.concat([old_log, log_entry], ignore_index=True)
    updated_log.to_excel(excel_file, index=False)

print("✅ Đã ghi log xử lý vào file Excel:", excel_file)



✅ Đã ghi log xử lý vào file Excel: ../reports/featureEngineering_summary.xlsx


In [4]:
df_train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave  None      Reg   
1   2          20       RL         80.0     9600   Pave  None      Reg   
2   3          60       RL         68.0    11250   Pave  None      IR1   
3   4          70       RL         60.0     9550   Pave  None      IR1   
4   5          60       RL         84.0    14260   Pave  None      IR1   
5   6          50       RL         85.0    14115   Pave  None      IR1   
6   7          20       RL         75.0    10084   Pave  None      Reg   
7   8          60       RL          NaN    10382   Pave  None      IR1   
8   9          50       RM         51.0     6120   Pave  None      Reg   
9  10         190       RL         50.0     7420   Pave  None      Reg   

  LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub  ...        0   None   None        None       0   
1         Lvl    AllPub  ...        0   None   None        None       0   
2         Lvl    AllPub  ...        0   None   None        None       0   
3         Lvl    AllPub  ...        0   None   None        None       0   
4         Lvl    AllPub  ...        0   None   None        None       0   
5         Lvl    AllPub  ...        0   None  MnPrv        Shed     700   
6         Lvl    AllPub  ...        0   None   None        None       0   
7         Lvl    AllPub  ...        0   None   None        Shed     350   
8         Lvl    AllPub  ...        0   None   None        None       0   
9         Lvl    AllPub  ...        0   None   None        None       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

## Các cột số bị thiếu như MasVnrArea(Diện tích tường ốp gạch)

Ngta bỏ trống chắc điền 0 vào sẽ hợp lí hơn

In [30]:
df_train['MasVnrArea'] = df_train['MasVnrArea'].fillna(0)
df_train['GarageYrBlt'] = df_train['GarageYrBlt'].fillna(0)

df_test['MasVnrArea'] = df_test['MasVnrArea'].fillna(0)
df_test['GarageYrBlt'] = df_test['GarageYrBlt'].fillna(0)
# Thêm kết luận quan trọng vào log
important_note = "Đã điền giá trị thiếu cho MasVnrArea và GarageYrBlt bằng 0. Đây là các cột số, giá trị thiếu hợp lý nhất là 0."
important_entry = pd.DataFrame({
    "Thời gian": [datetime.now().strftime("%Y-%m-%d %H:%M:%S")],
    "Hành động": ["Điền Missing Values bằng 0 cho MasVnrArea, GarageYrBlt"],
    "Kết luận": [important_note]
})

old_log = pd.read_excel(excel_file)
updated_log = pd.concat([old_log, important_entry], ignore_index=True)
updated_log.to_excel(excel_file, index=False)
print("✅ Đã ghi thêm kết luận quan trọng vào file Excel:", excel_file)

✅ Đã ghi thêm kết luận quan trọng vào file Excel: ../reports/featureEngineering_summary.xlsx


In [6]:
df_train['MasVnrArea'].isnull().sum()
df_train['MasVnrArea'].head()

0    196.0
1      0.0
2    162.0
3      0.0
4    350.0
Name: MasVnrArea, dtype: float64

## Điền giá trị phổ biến cho các cột thiếu ít

In [32]:
from datetime import datetime

# Thêm log cho xử lý Electrical
note_electrical = "Đã điền giá trị thiếu cho Electrical bằng giá trị mode (phổ biến nhất)."
log_electrical = pd.DataFrame({
    "Thời gian": [datetime.now().strftime("%Y-%m-%d %H:%M:%S")],
    "Hành động": ["Điền Missing Values cho Electrical bằng mode"],
    "Kết luận": [note_electrical]
})

old_log = pd.read_excel(excel_file)
updated_log = pd.concat([old_log, log_electrical], ignore_index=True)
updated_log.to_excel(excel_file, index=False)
print("✅ Đã ghi log xử lý Electrical vào file Excel:", excel_file)


✅ Đã ghi log xử lý Electrical vào file Excel: ../reports/featureEngineering_summary.xlsx


## Kiểm tra lại thử

In [ ]:
print(df_train.isnull().sum().sum())
print(df_test.isnull().sum().sum())

259
249


In [ ]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          1460 non-null   object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

Vừa khớp 259 giá trị thiếu ở LotFrontage 

## Điền giá trị LotFrontage bị thiếu bằng median theo từng Neighborhood

In [33]:
from datetime import datetime

# Tạo kết luận cho việc xử lý LotFrontage
note_lotfrontage = "Đã điền giá trị thiếu cho LotFrontage bằng median theo từng Neighborhood."
log_lotfrontage = pd.DataFrame({
    "Thời gian": [datetime.now().strftime("%Y-%m-%d %H:%M:%S")],
    "Hành động": ["Điền Missing Values cho LotFrontage bằng median theo Neighborhood"],
    "Kết luận": [note_lotfrontage]
})

old_log = pd.read_excel(excel_file)
updated_log = pd.concat([old_log, log_lotfrontage], ignore_index=True)
updated_log.to_excel(excel_file, index=False)
print("✅ Đã ghi log xử lý LotFrontage vào file Excel:", excel_file)

✅ Đã ghi log xử lý LotFrontage vào file Excel: ../reports/featureEngineering_summary.xlsx


In [11]:
df_train['LotFrontage'].isnull().sum()
df_test['LotFrontage'].isnull().sum()

0

## Mã hóa (Encoding) các biến phân loại (object)

Kiểm tra coi các cột của tập test và tập train có trùng với nhau không

In [34]:
from datetime import datetime

# Tạo kết luận và thời gian
important_note = "Kết luận quan trọng: Đã chuẩn hóa dữ liệu và xử lý các biến đặc trưng mới."
now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

# Tạo entry mới
important_entry = pd.DataFrame({
    "Thời gian": [now],
    "Hành động": ["Ghi kết luận quan trọng"],
    "Kết luận": [important_note]
})

# Đọc log cũ và ghi thêm dòng mới
old_log = pd.read_excel(excel_file)
updated_log = pd.concat([old_log, important_entry], ignore_index=True)
updated_log.to_excel(excel_file, index=False)
print("✅ Đã ghi kết luận quan trọng vào file Excel:", excel_file)

✅ Đã ghi kết luận quan trọng vào file Excel: ../reports/featureEngineering_summary.xlsx


Các cột của 2 tập trùng nhau 

## One hot encoding 

In [35]:
from datetime import datetime

# Tạo kết luận và thời gian
note_encoding = "Đã thực hiện one-hot encoding cho các biến phân loại (object) bằng pd.get_dummies."
now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

# Tạo entry mới
encoding_entry = pd.DataFrame({
    "Thời gian": [now],
    "Hành động": ["One-hot encoding với pd.get_dummies"],
    "Kết luận": [note_encoding]
})

# Đọc log cũ và ghi thêm dòng mới
old_log = pd.read_excel(excel_file)
updated_log = pd.concat([old_log, encoding_entry], ignore_index=True)
updated_log.to_excel(excel_file, index=False)
print("✅ Đã ghi kết luận one-hot encoding vào file Excel:", excel_file)


✅ Đã ghi kết luận one-hot encoding vào file Excel: ../reports/featureEngineering_summary.xlsx


In [17]:
df_train.head()

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          60         65.0     8450            7            5       2003   
1   2          20         80.0     9600            6            8       1976   
2   3          60         68.0    11250            7            5       2001   
3   4          70         60.0     9550            7            5       1915   
4   5          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLI  SaleType_ConLw  \
0          2003       196.0         706  ...           False           False   
1          1976         0.0         978  ...           False           False   
2          2002       162.0         486  ...           False           False   
3          1970         0.0         216  ...           False           False   
4          2000       350.0         655  ...           False           False   

   SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_AdjLand  \
0         False         False         True                  False   
1         False         False         True                  False   
2         False         False         True                  False   
3         False         False         True                  False   
4         False         False         True                  False   

   SaleCondition_Alloca  SaleCondition_Family  SaleCondition_Normal  \
0                 False                 False                  True   
1                 False                 False                  True   
2                 False                 False                  True   
3                 False                 False                 False   
4                 False                 False                  True   

   SaleCondition_Partial  
0                  False  
1                  False  
2                  False  
3                  False  
4                  False  

[5 rows x 261 columns]

## Tạo đặc trưng mới (Feature Engineering)

vd như tổng diện tích được cộng lại từ tầng hầm, tầng 1, tầng 2. Giá nhà (SalePrice) thường tăng mạnh theo diện tích sử dụng thực tế, chứ không chỉ từng phần riêng lẻ.

Mô hình dễ hiểu hơn khi có 1 cột tổng hợp thay vì phải học mối quan hệ rời rạc giữa 3 cột.

In [36]:
from datetime import datetime

# Tạo kết luận và thời gian
note_feature = "Đã tạo các đặc trưng mới: TotalSF, TotalBathrooms, Age cho cả train và test."
now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

# Tạo entry mới
feature_entry = pd.DataFrame({
    "Thời gian": [now],
    "Hành động": ["Tạo đặc trưng mới (TotalSF, TotalBathrooms, Age)"],
    "Kết luận": [note_feature]
})

# Đọc log cũ và ghi thêm dòng mới
old_log = pd.read_excel(excel_file)
updated_log = pd.concat([old_log, feature_entry], ignore_index=True)
updated_log.to_excel(excel_file, index=False)
print("✅ Đã ghi kết luận đặc trưng mới vào file Excel:", excel_file)


✅ Đã ghi kết luận đặc trưng mới vào file Excel: ../reports/featureEngineering_summary.xlsx


In [19]:
df_train['TotalBathrooms'].head()

0    3.5
1    2.5
2    3.5
3    2.0
4    3.5
Name: TotalBathrooms, dtype: float64

## Xử lý phân phối lệch cho biến mục tiêu SalePrice, lệch phải

In [38]:
from datetime import datetime

# Tạo kết luận và thời gian
note_saleprice = "Đã xử lý phân phối lệch phải cho biến mục tiêu SalePrice bằng log1p."
now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

# Tạo entry mới
saleprice_entry = pd.DataFrame({
    "Thời gian": [now],
    "Hành động": ["Xử lý SalePrice bằng log1p"],
    "Kết luận": [note_saleprice]
})

# Đọc log cũ và ghi thêm dòng mới
old_log = pd.read_excel(excel_file)
updated_log = pd.concat([old_log, saleprice_entry], ignore_index=True)
updated_log.to_excel(excel_file, index=False)
print("✅ Đã ghi log xử lý SalePrice vào file Excel:", excel_file)


✅ Đã ghi log xử lý SalePrice vào file Excel: ../reports/featureEngineering_summary.xlsx


## Chuẩn hóa dữ liệu

In [39]:
# Thêm kết luận quan trọng về chuẩn hóa dữ liệu vào file log Excel
note_important = "Đã chuẩn hóa dữ liệu số bằng StandardScaler cho cả train và test."
action_important = "Chuẩn hóa dữ liệu với StandardScaler"
now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

important_entry = pd.DataFrame({
    "Thời gian": [now],
    "Hành động": [action_important],
    "Kết luận": [note_important]
})

old_log = pd.read_excel(excel_file)
updated_log = pd.concat([old_log, important_entry], ignore_index=True)
updated_log.to_excel(excel_file, index=False)
print("✅ Đã ghi kết luận chuẩn hóa dữ liệu vào file Excel:", excel_file)


✅ Đã ghi kết luận chuẩn hóa dữ liệu vào file Excel: ../reports/featureEngineering_summary.xlsx


In [22]:
df_train.head()

Id  MSSubClass  LotFrontage   LotArea  OverallQual  OverallCond  \
0 -1.730865    0.073375    -0.231877 -0.207142     0.651479    -0.517200   
1 -1.728492   -0.872563     0.437043 -0.091886    -0.071836     2.179628   
2 -1.726120    0.073375    -0.098093  0.073480     0.651479    -0.517200   
3 -1.723747    0.309859    -0.454850 -0.096897     0.651479    -0.517200   
4 -1.721374    0.073375     0.615421  0.375148     1.374795    -0.517200   

   YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_Oth  \
0   1.050994      0.878668    0.514104    0.575425  ...         False   
1   0.156734     -0.429577   -0.570750    1.171992  ...         False   
2   0.984752      0.830215    0.325915    0.092907  ...         False   
3  -1.863632     -0.720298   -0.570750   -0.499274  ...         False   
4   0.951632      0.733308    1.366489    0.463568  ...         False   

   SaleType_WD  SaleCondition_AdjLand  SaleCondition_Alloca  \
0         True                  False                 False   
1         True                  False                 False   
2         True                  False                 False   
3         True                  False                 False   
4         True                  False                 False   

   SaleCondition_Family  SaleCondition_Normal  SaleCondition_Partial  \
0                 False                  True                  False   
1                 False                  True                  False   
2                 False                  True                  False   
3                 False                 False                  False   
4                 False                  True                  False   

    TotalSF  TotalBathrooms       Age  
0 -0.001277        1.642256 -1.043259  
1 -0.052407        0.368581 -0.183465  
2  0.169157        1.642256 -0.977121  
3 -0.114493       -0.268257  1.800676  
4  0.944631        1.642256 -0.944052  

[5 rows x 264 columns]

## Lưu file csv, mấy cái đã feature engineering 

In [23]:
import os
os.makedirs("../data/03_processed", exist_ok=True)
df_train.to_csv("../data/03_processed/train_clean.csv", index=False)
df_test.to_csv("../data/03_processed/test_clean.csv", index=False)
print("Dữ liệu đã sẵn sàng cho mô hình (03_processed).")

Dữ liệu đã sẵn sàng cho mô hình (03_processed).
